Find.md

```diff
@@ -0,0 +1,202 @@
+cartesian indices and values (`findnz`), range of linear indices (`search*`)
+
+- **Return when not found**: shorter vector for all-at-once functions (`find`, `findin`,
+`findn`, `findfnz`) except `indexin` (which includes a `0` entry), `0` for functions
+returning a single index
+
+## Summary of Current Status
+
+The following table reorganizes existing methods which return linear indices based on the
+first two dimensions described above:
+- Whether to return all matches, only the next one, or only the previous one.
+- What values to look for.
+
+|  | nonzeros | test predicate `pred` | in collection `c` | equal to value `v` | sequence or regex `s` | extrema |
+| --- | --- | --- | --- | --- | --- | --- |
+| All at once | `find(A)` | `find(pred,A)` | `findin(A,c)` | `searchsorted(A,v)` | | `indmin(A)`/`indmax(A)` |
+| Next match | `findnext(A,1)` | `findnext(pred,A,1)` | | `findnext(A,v,1)` | `search(A,s,1)` | |
+| Previous match | `findprev(A,endof(A))` | `findprev(pred,A,endof(A))` | | `findprev(A,v,endof(A))` | `rsearch(A,s,endof(A))` | |
+
+Some functions do not fit into this table:
+- `findfirst` and `findlast` are special cases of `findnext` and `findprev`.
+- `searchsortedfirst` and `searchsortedlast` give each a part of the result of `searchsorted`.
+- `[r]searchindex` give part of the result from `[r]search`.
+- `findn` and `findnz` do not return linear indices.
+- `indexin` is similar to `findin` but returns `0` for entries with no match.
+- `match`, `matchall` and `eachmatch` return `RegexMatch` objects or strings rather than
+indices.
+- `findmin` and `findmax` return both the index an value of extrema.
+
+## Open Design Issues
+
+- **How to switch between forward and backward search**:
+    - Separate functions (e.g. starting with `r` for "reverse"): not great for
+    documentation, harder to find, not very Julian.
+    - `rev=false` positional argument: not very explicit.
+    - `rev=false` keyword argument: clearer and consistent with `sort`, but maybe too slow
+    (especially for single-element functions).
+    - special object like `Order.Forward`/`Order.Backward`: clearer, but these objects do
+    not have this meaning in Base, and introducing separate objects just for this may not
+    be worth it.
+
+- **Whether to keep the the `find`/`search` distinction**:
+    - Obviously requires clearly distinct meanings for each family of functions.
+    - Advantage: less complex signatures (due to many methods) for users, and limits
+    dispatch ambiguities.
+    - Drawback: using two different names for related functions makes it harder to find
+    one variant when you know the other one; in particular, auto-completion does not help.
+
+- **How to search iteratively**:
+    - Functions returning the next/previous match after a given index: simple, but require
+    manual handling of indices.
+    - Functions returning an iterator over matches: more user-friendly, but overkill when
```

Comment:
Technically, you can't pass an index as a state to `next`, since there is no guarantee that state and indices are the same thing (cf. https://github.com/JuliaLang/julia/pull/18668). But of course you can call `first` on the iterator, as long as it can be passed an index to start from (instead of always starting at `1`). I've added some discussion of this in the last commit.